### PR TITLE
added graphite plugin for data initialisation

### PR DESCRIPTION
**Pull request by karanshah03**
Added graphite plugin for data initialization.

### CODE REVIEW

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     * Client to load metric data from graphite.
+     */
+    @NonNull
+    private GraphiteClient graphiteClient;
+
+    @Override
+    public List<DataSourceResult> getMetricData(String from, Integer maxDataPoints, String metric) {
+        GraphiteResult[] graphiteResult = graphiteClient.getMetricData(from, maxDataPoints, metric);
+        List<DataSourceResult> results = new ArrayList<>();
+        if (graphiteResult.length != 0) {
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java]: Please add a test for a `graphiteResult` array with length 0.

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            String[][] dataPoints = graphiteResult[0].getDatapoints();
+            for (String[] dataPoint : dataPoints) {
+                Double dataPointValue = 0.0;
+                long epochSeconds = Long.parseLong(dataPoint[1]);
+                if (dataPoint[0] != null) {
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java]: Please add test case for null value.

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    public GraphiteResult[] getMetricData(String from, Integer maxDataPoints, String target) {
+        notNull(target, "target can't be null");
+        // https://graphite/render?from=-7d&format=json&maxDataPoints=2016&target=sortByMaxima(groupByNodes(stats.gauges.adaptive-alerting-metric-mapper.*.NORMALIZED.kafka-message-processor.consumed.count,%20%27sum%27,%200))
+        // https://graphite/render?from=-600s&until=now&format=json&target=seriesByTag("name=cpu")|aliasByTags("server","name")
+
+        val uri = String.format(baseUri + FETCH_METRICS_PATH, getValueOrDefault(from, "1d"), getValueOrDefault(maxDataPoints, 288), target);
+        Content content;
+        try {
+            content = httpClient.get(uri);
+        } catch (IOException e) {
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java]: Please add a test case that covers this exception handling block.

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Test
+    public void testGetMetricData() {
+        val result = sourceUnderTest.getMetricData("1d", 2016, "");
+        assertEquals(dataSourceResults, result);
+    }
+
+    private void initTestObjects() {
+        GraphiteResult graphiteResult = new GraphiteResult();
+        String[][] dataPoints = new String[2][2];
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Suggest converting this array to a constant

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        dataPoints[0][0] = "1";
+        dataPoints[0][1] = "1578307488";
+        dataPoints[1][0] = "3";
+        dataPoints[1][1] = "1578307489";
+        graphiteResult.setDatapoints(dataPoints);
+        graphiteResults[0] = graphiteResult;
+
+        DataSourceResult dataSourceResult = new DataSourceResult();
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Suggest extracting these 3 lines into a method.

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClientTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private Content docContent_cantRead;
+
+    private byte[] docsBytes = "docsBytes".getBytes();
+
+    private byte[] docBytes_cantGet = "docBytes_cantGet".getBytes();
+
+    private byte[] docBytes_cantRead = "docBytes_cantRead".getBytes();
+
+
+    private GraphiteResult[] docs = {};
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClientTest.java]: Suggest moving static fields up to top of class.

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClientTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    private Content docsContent;
+
+    @Mock
+    private Content docContent_cantGet;
+
+    @Mock
+    private Content docContent_cantRead;
+
+    private byte[] docsBytes = "docsBytes".getBytes();
+
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClientTest.java]: Suggest collapsing white space between these 3 fields.

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    private GraphiteClient graphiteClient;
+
+    @Override
+    public List<DataSourceResult> getMetricData(String from, Integer maxDataPoints, String metric) {
+        GraphiteResult[] graphiteResult = graphiteClient.getMetricData(from, maxDataPoints, metric);
+        List<DataSourceResult> results = new ArrayList<>();
+        if (graphiteResult.length != 0) {
+            String[][] dataPoints = graphiteResult[0].getDatapoints();
+            for (String[] dataPoint : dataPoints) {
+                Double dataPointValue = 0.0;
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java]: Using `0.0` here could be problematic as `0.0` implies that we received a valid value from Graphite. We should either default to a placeholder value (like `Double.NEGATIVE_INFINITY`) or remove this datapoint from the result altogether.

Stylisitically, suggest naming this variable `value` - the context makes it pretty clear it's related to datapoint.

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                long epochSeconds = Long.parseLong(dataPoint[1]);
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java]: Suggest moving this line to the bottom of the `for` block - just a stylistic comment... processing the 2nd item in the `dataPoint` array before the 1st item makes it awkward to read.

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+     * @param maxDataPoints max no of data points in result
+     * @param target        metric name or tag with an optional graphite function
+     * @return time series for the specified metric
+     */
+    public GraphiteResult[] getMetricData(String from, Integer maxDataPoints, String target) {
+        notNull(target, "target can't be null");
+        // https://graphite/render?from=-7d&format=json&maxDataPoints=2016&target=sortByMaxima(groupByNodes(stats.gauges.adaptive-alerting-metric-mapper.*.NORMALIZED.kafka-message-processor.consumed.count,%20%27sum%27,%200))
+        // https://graphite/render?from=-600s&until=now&format=json&target=seriesByTag("name=cpu")|aliasByTags("server","name")
+
+        val uri = String.format(baseUri + FETCH_METRICS_PATH, getValueOrDefault(from, "1d"), getValueOrDefault(maxDataPoints, 288), target);
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java]: Suggest extracting parameters out into easy to understand variables:

```
        String fromValue = getValueOrDefault(from, "1d");
        Integer maxDataPointsValue = getValueOrDefault(maxDataPoints, 288);
        val uri = String.format(baseUri + FETCH_METRICS_PATH, fromValue, maxDataPointsValue, target);
```

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+     * Fetch metric data for a given set of metrics
+     *
+     * @param from          earliest time
+     * @param maxDataPoints max no of data points in result
+     * @param target        metric name or tag with an optional graphite function
+     * @return time series for the specified metric
+     */
+    public GraphiteResult[] getMetricData(String from, Integer maxDataPoints, String target) {
+        notNull(target, "target can't be null");
+        // https://graphite/render?from=-7d&format=json&maxDataPoints=2016&target=sortByMaxima(groupByNodes(stats.gauges.adaptive-alerting-metric-mapper.*.NORMALIZED.kafka-message-processor.consumed.count,%20%27sum%27,%200))
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java]: Remove these 2 comment lines

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java (hunk 10 above)
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java]: Why did we choose `1d` and `288`?  
Should we make this default configurable?  Should we even have a default?
At the very least please document this choice and why it was made

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/DataSource.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package com.expedia.adaptivealerting.anomdetect.source.data;
+
+import java.util.List;
+
+/**
+ * Data source interface
+ * <p>
+ * Fetches metric data from the source for the specified metric.
+ * Currently it supports only graphite as data source.
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/DataSource.java]: This comment line is not required.  If/When it becomes invalidated (i.e. a new DataSource implementation is added) this comment is unlikely to be updated and will become confusing. Please remove.

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        this.sourceUnderTest = new GraphiteSource(graphiteClient);
+    }
+
+    @Test
+    public void testGetMetricData() {
+        val dataSourceResult = getDataSourceResult(1.0, 1578307488);
+        DataSourceResult dataSourceResult1 = getDataSourceResult(3.0, 1578307489);
+
+        List<DataSourceResult> dataSourceResults = new ArrayList<>();
+        dataSourceResults.add(dataSourceResult);
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Suggest inlining `dataSourceResult` and `dataSourceResult1` variables for readability:

```
        dataSourceResults.add(getDataSourceResult(1.0, 1578307488));
        dataSourceResults.add(getDataSourceResult(3.0, 1578307489));
```

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    private GraphiteResult getNullValueResult() {
+        GraphiteResult graphiteResult = new GraphiteResult();
+        String[][] dataPoints = new String[1][2];
+        dataPoints[0][0] = null;
+        dataPoints[0][1] = "1578307488";
+        graphiteResult.setDatapoints(dataPoints);
+        return graphiteResult;
+    }
+
+    private DataSourceResult getDataSourceResult(Double value, long epochSecs) {
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Suggest renaming this to `builtDataSourceResult` as `getDataSourceResult` implies the test is retrieving it from somewhere.

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        graphiteResults_null[0] = getNullValueResult();
+    }
+
+    private void initDependencies() {
+        when(graphiteClient.getMetricData(DEFAULT_FROM_VALUE, DEFAULT_MAX_DATA_POINTS, "metric_name")).thenReturn(graphiteResults);
+        when(graphiteClient.getMetricData(DEFAULT_FROM_VALUE, DEFAULT_MAX_DATA_POINTS, "null_metric")).thenReturn(new GraphiteResult[0]);
+        when(graphiteClient.getMetricData(DEFAULT_FROM_VALUE, DEFAULT_MAX_DATA_POINTS, "null_value")).thenReturn(graphiteResults_null);
+    }
+
+    private GraphiteResult getResult() {
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Suggest renaming this to `buildGraphiteResult`

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        String[][] dataPoints = new String[2][2];
+        dataPoints[0][0] = "1";
+        dataPoints[0][1] = "1578307488";
+        dataPoints[1][0] = "3";
+        dataPoints[1][1] = "1578307489";
+        graphiteResult.setDatapoints(dataPoints);
+        return graphiteResult;
+    }
+
+    private GraphiteResult getNullValueResult() {
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: `buildNullValueGraphiteResult`

File: anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    @Test
+    public void testGetMetricData_null_metric_data() {
+        val actual = sourceUnderTest.getMetricData(DEFAULT_FROM_VALUE, DEFAULT_MAX_DATA_POINTS, "null_metric");
+        assertEquals(new ArrayList<>(), actual);
+    }
+
+    @Test
+    public void testGetMetricData_null_value() {
+        val actual = sourceUnderTest.getMetricData(DEFAULT_FROM_VALUE, DEFAULT_MAX_DATA_POINTS, "null_value");
+        val dataSourceResult = getDataSourceResult(Double.NEGATIVE_INFINITY, 1578307488);
```
Comment (mattcallanan) [inline anomdetect/src/test/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSourceTest.java]: Use a constant e.g. `GraphiteSource.MISSING_VALUE`.  See GraphiteSource.java review comment

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    private GraphiteClient graphiteClient;
+
+    @Override
+    public List<DataSourceResult> getMetricData(String from, Integer maxDataPoints, String metric) {
+        GraphiteResult[] graphiteResult = graphiteClient.getMetricData(from, maxDataPoints, metric);
+        List<DataSourceResult> results = new ArrayList<>();
+        if (graphiteResult.length != 0) {
+            String[][] dataPoints = graphiteResult[0].getDatapoints();
+            for (String[] dataPoint : dataPoints) {
+                Double value = Double.NEGATIVE_INFINITY;
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteSource.java]: This should be exposed as a constant value, e.g. `MISSING_VALUE`.  
Test case should not rely on `Double.NEGATIVE_INFINITY`

File: anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+     * Fetch metric data for a given set of metrics
+     *
+     * @param from          earliest time
+     * @param maxDataPoints max no of data points in result
+     * @param target        metric name or tag with an optional graphite function
+     * @return time series for the specified metric
+     */
+    public GraphiteResult[] getMetricData(String from, Integer maxDataPoints, String target) {
+        notNull(target, "target can't be null");
+        val fromValue = getValueOrDefault(from, DEFAULT_FROM_VALUE);
```
Comment (mattcallanan) [inline anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/source/data/graphite/GraphiteClient.java]: I'm still not convinced that these are valid default values to apply globally. 
I think it would be better to have `notNull` checks in this method for `from` and `maxDataPoints` params.  
Also, thanks for removing the Expedia-specific comment.  But it would be good to have a (generic) example of what "target" can be in the JavaDoc for this method.
